src/programs/simulate/simulate.cpp

```diff
@@ -3126,26 +3210,33 @@ void SimulateApp::fill_water_potential(const PDB* current_specimen, Image* scatt
     projected_water_atoms.Allocate(scattering_slab->logical_x_dimension, scattering_slab->logical_y_dimension, 1);
     projected_water_atoms.SetToConstant(0.0f);
 
-    // To compare the thread block ordering, undo with schedule(dynamic,1)
-    // schedule(static, water_box->number_of_waters /number_of_threads )
+    float const* transformed_water_pos = water_box->GetTransformedPositions( );
 
-    //    timer.lap("water_pre");
+    float slab_min_z = (float)slabIDX_start[iSlab];
+    float slab_max_z = (float)slabIDX_end[iSlab] + 1;
 
-    long n_waters_ignored = 0;
-#pragma omp parallel for num_threads(this->number_of_threads)                                                                                                        \
-        schedule(static, water_box->number_of_waters / number_of_threads) private(radius, ix, iy, iz, dx, dy, dz, x1, y1, z1, indX, indY, indZ, int_x, int_y, int_z, \
-                                                                                  sx, sy, iSubPixX, iSubPixY, iSubPixZ, iSubPixLinearIndex,                          \
-                                                                                  n_waters_ignored, current_weight, current_distance, current_potential)
+    auto slab_indices = water_box->GetSlabRange(slab_min_z - rotated_oZ - pixel_offset, slab_max_z - rotated_oZ - pixel_offset);
 
-    for ( int current_atom = 0; current_atom < water_box->number_of_waters; current_atom++ ) {
+    for ( int current_atom = slab_indices.first; current_atom < slab_indices.second; current_atom++ ) {
```

Comment:
We leverage the fact that the water box is sorted to quickly compute the range of water atoms we should process in each slab. No need to test those that are too far away. This is always a contiguous segment in the array of water atoms.